modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java

```diff
@@ -866,11 +866,17 @@ protected Experiment getExperiment(Application.Name applicationName, Experiment.
         Experiment result = null;
 
         if (metadataCacheEnabled) {
-            List<Experiment> experiments = metadataCache.getExperimentsByAppName(applicationName);
-            for (Experiment exp : experiments) {
-                if (experimentLabel.equals(exp.getLabel())) {
-                    result = exp;
-                    break;
+            //First fetch experiment list sorted by priorities (contains non-deleted experiment-ids).
+            //This experiment-priority list is the source of truth for ALL assignment flows while looking up experiment...
+            Optional<PrioritizedExperimentList> prioritizedExperimentListOptional = metadataCache.getPrioritizedExperimentListMap(applicationName);
+            if (prioritizedExperimentListOptional.isPresent()) {
+                //Iterate as per experiment priority and look for the matching experiment by their label
+                for (PrioritizedExperiment prioritizedExperiment : prioritizedExperimentListOptional.get().getPrioritizedExperiments()) {
+                    if (experimentLabel.equals(prioritizedExperiment.getLabel())) {
+                        //Upon match, get the complete experiment object from cache
+                        result = metadataCache.getExperimentById(prioritizedExperiment.getID()).orElseGet(null);
+                        break;
+                    }
                 }
             }
         } else {
```

Comment:
@bargenilesh Food for thought: The implementation for `getExperiment` made me think that why not also have a cache on top of the `experiment_label_index` table which is the source of this method call. It will help in de-cluttering a lot of iteration logic that we need to put together when getting the corresponding experiment from the priority list. Maybe something that we could add as a TODO/tech debt?